apps/typegpu-docs/src/content/examples/rendering/3d-fish/index.ts

```diff
@@ -0,0 +1,414 @@
+});
+
+const computePipeline = root['~unstable']
+  .withCompute(mainCompute)
+  .createPipeline();
+
+// bind groups
+
+const sampler = root.device.createSampler({
+  addressModeU: 'repeat',
+  addressModeV: 'repeat',
+  magFilter: 'linear',
+  minFilter: 'linear',
+});
+
+const renderFishBindGroups = [0, 1].map((idx) =>
+  root.createBindGroup(renderBindGroupLayout, {
+    modelData: fishDataBuffers[idx],
+    camera: cameraBuffer,
+    modelTexture: fishModel.texture,
+    sampler: sampler,
+  }),
+);
+
+const renderOceanFloorBindGroups = root.createBindGroup(renderBindGroupLayout, {
+  modelData: oceanFloorDataBuffer,
+  camera: cameraBuffer,
+  modelTexture: oceanFloorModel.texture,
+  sampler: sampler,
+});
+
+const computeBindGroups = [0, 1].map((idx) =>
+  root.createBindGroup(computeBindGroupLayout, {
+    currentFishData: fishDataBuffers[idx],
+    nextFishData: fishDataBuffers[1 - idx],
+    mouseRay: mouseRayBuffer,
+    timePassed: timePassedBuffer,
+  }),
+);
+
+// frame
+
+let odd = false;
+let lastTimestamp = Date.now();
+let disposed = false;
+
+function frame() {
+  if (disposed) {
+    return;
+  }
+  odd = !odd;
+
+  const currentTimestamp = Date.now();
+  timePassedBuffer.write(currentTimestamp - lastTimestamp);
+  lastTimestamp = currentTimestamp;
+  cameraBuffer.write(camera);
+  mouseRayBuffer.write(mouseRay);
+
+  computePipeline
+    .with(computeBindGroupLayout, computeBindGroups[odd ? 1 : 0])
+    .dispatchWorkgroups(p.fishAmount / p.workGroupSize);
+
+  renderPipeline
+    .withColorAttachment({
+      view: context.getCurrentTexture().createView(),
+      clearValue: [
+        p.backgroundColor.x,
+        p.backgroundColor.y,
+        p.backgroundColor.z,
+        1,
+      ],
+      loadOp: 'clear' as const,
+      storeOp: 'store' as const,
+    })
+    .withDepthStencilAttachment({
+      view: depthTexture.createView(),
+      depthClearValue: 1,
+      depthLoadOp: 'clear',
+      depthStoreOp: 'store',
+    })
+    .with(modelVertexLayout, oceanFloorModel.vertexBuffer)
+    .with(renderInstanceLayout, oceanFloorDataBuffer)
+    .with(renderBindGroupLayout, renderOceanFloorBindGroups)
+    .draw(oceanFloorModel.polygonCount, 1);
+
+  renderPipeline
+    .withColorAttachment({
+      view: context.getCurrentTexture().createView(),
+      clearValue: [
+        p.backgroundColor.x,
+        p.backgroundColor.y,
+        p.backgroundColor.z,
+        1,
+      ],
+      loadOp: 'load' as const,
+      storeOp: 'store' as const,
+    })
+    .withDepthStencilAttachment({
+      view: depthTexture.createView(),
+      depthClearValue: 1,
+      depthLoadOp: 'load',
+      depthStoreOp: 'store',
+    })
+    .with(modelVertexLayout, fishModel.vertexBuffer)
+    .with(renderInstanceLayout, fishDataBuffers[odd ? 1 : 0])
+    .with(renderBindGroupLayout, renderFishBindGroups[odd ? 1 : 0])
+    .draw(fishModel.polygonCount, p.fishAmount);
+
+  root['~unstable'].flush();
+
+  requestAnimationFrame(frame);
+}
+frame();
```

Comment:
(After changing callback)
```suggestion
requestAnimationFrame(frame);
```